### DISCUSSION TITLE
POSTPONE a deferred word.

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#184] [Q&A] [answered]
What does it mean to `POSTPONE` a deferred word?  Is it ambiguous or is there some historical undocumented behaviour, like use the value of the deferred word at the time `POSTPONE` are applied?

(hypothetical example) 
```forth
DEFER return IMMEDIATE

: alt_exit ( -- )
    ['] xyzzy ['] return DEFER!
    ...
    POSTPONE return
    ['] EXIT ['] return DEFER!
; IMMEDIATE

: woot ( x -- )
    ?DUP IF alt_exit THEN
    ...
;
```

So what happens when `0 woot` executes?  Does it follow `EXIT` or `xyzzy`?

Does something similar happen for `COMPILE,`?

**Comment by ruv** [comment, marked as answer]
> What does it mean to `POSTPONE` a deferred word?

Nothing special. Same as for other user-defined words:
  - if a deferred word `foo` is an ordinary word, then `postpone foo` is equivalent to `['] foo compile,`;
  - if a deferred word `foo` is an immediate word, then `postpone foo` is equivalent to `['] foo execute-compiling`.

-----

> use the value of the deferred word at the time `POSTPONE` are applied?

Do you mean the compilation semantics of `postpone` or the run-time semantics that they appends?

> (hypothetical example)

A standard program cannot tick `exit` (or, at least, execute the returned _xt_). Because it is not allowed to obtain *the execution token* of a word for which interpretation semantics are not defined by the standard. You can use `[: POSTPONE EXIT ;]` instead of  `['] EXIT` .

Because `return` is an immediate word, your definition for `alt_exit` is equivalent to the following:
```forth
: alt_exit ( -- )
    ['] xyzzy ['] return DEFER!
    ...
    ['] return execute-compiling
    [: POSTPONE EXIT ;] ['] return DEFER!
; IMMEDIATE

: woot ( x -- )
    ?DUP IF alt_exit THEN
    ...
;
```
So, the execution semantics of `return` are preformed before they are changed in the second time.
Hence, this definition for `alt_exit` is equivalent to the following:
```forth
: alt_exit ( -- )
    ['] xyzzy ['] return DEFER!
    ...
    ['] xyzzy execute-compiling
    [: POSTPONE EXIT ;] ['] return DEFER!
; IMMEDIATE
```

> So what happens when `0 woot` executes? Does it follow `EXIT` or `xyzzy`?

It follows `xyzzy`.

**Reply by ruv** [reply]
>>    So what happens when `0 woot` executes? Does it follow EXIT or xyzzy?
>>
> It follows `xyzzy`.

```forth
: woot ( x -- )
    ( x.some\0 | 0 ) ?DUP ( 0 | x.some x.some ) IF ( x.some ) alt_exit THEN
    ...
;
```

In fact, on `0 woot`  the run-time semantics at the point where `alt_exit` is used will not get control.  They will get control on `1 woot`.

**Reply by SirWumpus** [reply]
(Ugh. `1 woot` yep.  I blame the puppy for making me rush proof reading the question.  Bad puppy🐶)
